tests/unit/test_orchestrate.py

```diff
@@ -0,0 +1,378 @@
+    def test_no_failures():
+        test_calls = []
+
+        def make_test(name):
+            def test():  # pragma: NO COVER
+                test_calls.append(name)  # pragma: SYNCPOINT
+                test_calls.append(name)  # pragma: SYNCPOINT
+                test_calls.append(name)
+
+            return test
+
+        test1 = make_test("A")
+        test2 = make_test("B")
+
+        permutations = orchestrate._permutations(["A", "B", "A", "B", "A", "B"])
+        expected = list(itertools.chain(*permutations))
+
+        counts = orchestrate.orchestrate(test1, test2)
+        assert counts == (3, 3)
+        assert test_calls == expected
+
+    @staticmethod
+    def test_named_syncpoints():
+        test_calls = []
+
+        def make_test(name):
+            def test():  # pragma: NO COVER
+                test_calls.append(name)  # pragma: SYNCPOINT test_named_syncpoints
+                test_calls.append(name)  # pragma: SYNCPOINT test_named_syncpoints
+                test_calls.append(name)  # pragma: SYNCPOINT
+
+            return test
+
+        test1 = make_test("A")
+        test2 = make_test("B")
+
+        permutations = orchestrate._permutations(["A", "B", "A", "B", "A", "B"])
+        expected = list(itertools.chain(*permutations))
+
+        counts = orchestrate.orchestrate(test1, test2, name="test_named_syncpoints")
+        assert counts == (3, 3)
+        assert test_calls == expected
+
+    @staticmethod
+    def test_syncpoints_decrease_after_initial_run():
+        test_calls = []
+
+        def make_test(name):
+            syncpoints = [name] * 4
+
+            def test():  # pragma: NO COVER
+                test_calls.append(name)
+                if syncpoints:
+                    syncpoints.pop()  # pragma: SYNCPOINT
+                    test_calls.append(name)
+
+            return test
+
+        test1 = make_test("A")
+        test2 = make_test("B")
+
+        expected = [
+            "A",
+            "A",
+            "B",
+            "B",
+            "A",
+            "B",
+            "A",
+            "B",
+            "A",
+            "B",
+            "B",
+            "A",
+            "B",
+            "A",
+            "A",
+            "B",
+            "B",
+            "A",
+            "B",
+            "A",
+        ]
+
+        counts = orchestrate.orchestrate(test1, test2)
+        assert counts == (2, 2)
+        assert test_calls == expected
+
+    @staticmethod
+    def test_syncpoints_increase_after_initial_run():
+        test_calls = []
+
+        def do_nothing():  # pragma: NO COVER
+            pass
+
+        def make_test(name):
+            syncpoints = [None] * 4
+
+            def test():  # pragma: NO COVER
+                test_calls.append(name)  # pragma: SYNCPOINT
+                test_calls.append(name)
+
+                if syncpoints:
+                    syncpoints.pop()
+                else:
+                    do_nothing()  # pragma: SYNCPOINT
```

Comment:
This shows a limitation of the current `settrace` approach, in that the comment must be on a reachable line, so the call to `do_nothing` is just here to have something to hang the pragma off of. It would be more satisfying to be able to write something like:

~~~~
if syncpoints:
    syncpoints.pop()
else:
    # pragma: SYNCPOINT
    test_calls.append(name)
~~~~

I spent some time trying to figure out how to make this possible and concluded it was complicated enough to leave out for the first pass. Unfortunately, neither `tokenize` nor `ast`, on their own, provide enough information to figure out which side of the `else` the pragma is on and assign to a line of code. It should be possible to use both to figure it out, but it's not trivial.

Reply:
I think it's fine for the pragma to mark the syncpoint.